### PR TITLE
Codegen overview doc

### PR DESCRIPTION
**Pull request by alex-tpom6oh**

### CODE REVIEW

**Comment by hugihlynsson** [comment]
#### 📝 Changed routes:


| [**…/start-using-inspector**](https://docs-hfgnx5oaz.avo-dev.app/data-design/start-using-inspector) | [**…n/avo-codegen-overview**](https://docs-hfgnx5oaz.avo-dev.app/implementation/avo-codegen-overview) |
|:-:|:-:|
|<a href="https://docs-hfgnx5oaz.avo-dev.app/data-design/start-using-inspector">Screenshot of /data-design/start-using-inspector</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/biW4bZA2bF35TTxxrTQC_gTK_J-aaL2t93ttO0H5ukskkprBZnUIqPlsH3fu5G9Pr1txYoMEcEvxtlMfdwYXEg.png">(view full size)</a> |<a href="https://docs-hfgnx5oaz.avo-dev.app/implementation/avo-codegen-overview">Screenshot of /implementation/avo-codegen-overview</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/nSt0uvnAchuIUKLOIPv7EkuLoWMI9kN1peJhXZqFMRRJ6TG1MwXWM0XJKKcET9NgibjHJw1LWHgIW6Ifnm1v4A.png">(view full size)</a> |



| [**…tart-using-avo-codegen**](https://docs-hfgnx5oaz.avo-dev.app/implementation/start-using-avo-codegen) | [**/workflow/implement**](https://docs-hfgnx5oaz.avo-dev.app/workflow/implement) |
|:-:|:-:|
|<a href="https://docs-hfgnx5oaz.avo-dev.app/implementation/start-using-avo-codegen">Screenshot of /implementation/start-using-avo-codegen</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/g5gkps5KGUA0CVNerWJBzTCpwlmOVOYULeux5_9906MP9NIkR2wrPHu-UYkic7RQNfTtd5sP8gm9XrHiBxDjCA.png">(view full size)</a> |<a href="https://docs-hfgnx5oaz.avo-dev.app/workflow/implement">Screenshot of /workflow/implement</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/LZgBIFQeE5V-QW_XYMnWbAy3TblZRINd03rs3yZgkRtDE6F6Z9EkJ47AXIMCnjg8eETppHOvPzIU3UxhSCGqKA.png">(view full size)</a> |





1 deleted route:
- **/implementation/devs-101**

Commit 0b01cc413a1475ed8ae4811a2f9bd913f2417cec (https://docs-hfgnx5oaz.avo-dev.app).

**Review by hugihlynsson — APPROVED** [review]
Did a quick pass and it looks great!

File: pages/data-design/start-using-inspector.mdx [hunk 1]
```diff
@@ -25,6 +25,12 @@ Inspector gives you:
 - Implementation status for the events in your entire tracking plan, so you can see which events are actually implemented and whether they’re sent correctly or not.
 - Inspector helps you migrate your tracking plan to Avo. Skip importing things you’re skeptical of. Then leverage Inspector to detect how events and properties are actually tracked, and gradually add them to your tracking plan via Avo branches.
 
+### How does Inspector work
+
+Inspector API extracts the schemas (aka shapes) of the events you trigger (no PII data,
+only event names, property names, and types of properties) and sends those schemas to
+the Avo servers. To repeat: None of the PII data ever flows through Avo servers; only event schemas.
```
Comment (logason) [inline pages/data-design/start-using-inspector.mdx:32]: Removing API here to prevent the misunderstanding of the actual data hitting the API

```suggestion
Inspector API extracts the schemas (aka shapes/metadata) of the events you trigger. Inspector does not process property values or PII data, only event names, property names, and types of properties, and sends those schemas to the Avo servers.

TL;DR: No PII data ever flows through Avo servers; only event schemas.
```

File: pages/data-design/start-using-inspector.mdx [hunk 2]
```diff
@@ -72,7 +78,7 @@ Find the tag snippet and instructions to send data <Link passHref href="/impleme
 
 ### Install Inspector SDK
 
-If you’re not using one of the supported CDPs you can send events data with an SDK. It will require one time setup and a new app release and takes 2-30 minutes, depending on the developer’s experience. We recommend asking a developer that is familiar with the existing event tracking setup to help with this setup.
+If you’re not using one of the supported CDPs you can send events data with an SDK. It will require one time setup and a new app release and takes 10-30 minutes, depending on the developer’s experience. We recommend asking a developer that is familiar with the existing event tracking setup to help with this setup.
```
Comment (logason) [inline pages/data-design/start-using-inspector.mdx:81]: This paragraph indicates that you should use the SDK if available, rather than a CDP.

I'm not sure that's the messaging we want to go for here. I'd suggest flipping it around. If you are using one of our CDPs you should go for that integration, because it's quicker and doesn't require developer buy-in!

File: pages/data-design/start-using-inspector.mdx [hunk 3]
```diff
@@ -100,6 +106,13 @@ Learn how to set up SDKs for supported platforms:
 
 Not seeing your platform in the list? <Link passHref href="/help/troubleshooting"><a>Contact us</a></Link> if you want to send data from another platform.
 
+### Does Inspector receive PII data?
+
+No. Inspector does not receive any PII data. Inspector only takes in event names,
+property names, and types of properties.
+
+Learn more technical details about the <Link passHref href="/implementation/avo-inspector-overview"><a>Avo Inspector here</a></Link>.
```
Comment (logason) [inline pages/data-design/start-using-inspector.mdx:114]:
```suggestion
No. Inspector API does not receive any PII data. Inspector only takes schema metadata, containing event names, property names, and types of properties.

Learn more technical details about the <Link passHref href="/implementation/avo-inspector-overview"><a>Avo Inspector here</a></Link>.
```

File: pages/implementation/devs-101.mdx [hunk 4]
```diff
@@ -1,152 +0,0 @@
----
```
Comment (logason) [inline pages/implementation/devs-101.mdx:1]: Huge! Didn't know we were deprecating this one!

Have we made sure we're still covering a high level overview of all our dev offerings somewhere, and how they compare (with links to deep-dive materials)? Have we made sure to setup a redirect from this page to that one?

May gut says we should keep this page, but boil it down to more high level, and link to the new juicy docs we've been building for Inspector/Codegen for the details.

Would that make sense?

File: pages/implementation/start-using-avo-codegen.mdx [hunk 5]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+
+Codegen simplifies implementation and makes sure it is correct and consistent across all platforms.
+
+For example, instead of Android calling `analytics.track("game started")` and iOS calling `analytics.track("gameStarted")` , they both call `Avo.gameStarted()`, and the Avo function takes care of the spelling of the event and property names as they get passed into the analytics SDK.
+
+In other words, event and property naming is abstracted entirely away from the event implementation layer, ensuring that events and properties are named the same across all platforms, teams and code bases.
+
+## When use Codegen
+
+Codegen is our take on increasing the robustness of the analytics implementation and simplifying the implementation process.
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:25]:
```suggestion
Codegen is our way of increasing the robustness of the analytics implementation and simplifying the implementation process.
```

File: pages/implementation/start-using-avo-codegen.mdx [hunk 6]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+
+Adopting Codegen is not required to get value from Avo - many companies have their own ways to achieve robust analytics implementation and prefer to stick to them after adopting Avo and it’s totally fine.
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:27]:
```suggestion
Adopting Codegen is not required to get value from Avo - many companies have their own ways to achieve robust analytics implementation and prefer to stick to them rather than adopting a new approach.
```

File: pages/implementation/start-using-avo-codegen.mdx [hunk 7]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+
+Avo is designed to be efficient and powerful without the Codegen, providing the option of the advanced tracking plan management with any implementation approach.
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:29]: Maybe add links to tracking plan and inspector here?
```suggestion
Avo is designed to be efficient and powerful without the Codegen, providing advanced tracking plan management and observability with Inspector. 
```

File: pages/implementation/start-using-avo-codegen.mdx [hunk 8]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+In other words, event and property naming is abstracted entirely away from the event implementation layer, ensuring that events and properties are named the same across all platforms, teams and code bases.
+
+## When use Codegen
+
+Codegen is our take on increasing the robustness of the analytics implementation and simplifying the implementation process.
+
+Adopting Codegen is not required to get value from Avo - many companies have their own ways to achieve robust analytics implementation and prefer to stick to them after adopting Avo and it’s totally fine.
+
+Avo is designed to be efficient and powerful without the Codegen, providing the option of the advanced tracking plan management with any implementation approach. 
+The decision to use Codegen is usually in the hands of the Engineering department. If they prefer to use a different approach Avo would help you communicate the tracking plan changes to the engineering team and help them to give the implementation feedback.
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:30]: Want to avoid using "them" as a term for engineers here. I think we should be writing this doc for developers, who perhaps received a link from a data person.
```suggestion
The decision to use Codegen is usually in the hands of the engineering team on each platform. If you prefer to use a different approach for implementing your tracking code Avo can still help you by providing a platform to collaborate and align on the tracking specs, and by providing clear specs for the suggested tracking changes.
```

File: pages/implementation/start-using-avo-codegen.mdx [hunk 9]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+
+## How does Codegen work
+
+Codegen works with any analytics tool, whether it's any of the third-party event logging SDKs (e.g. Google Analytics 4, Segment, RudderStack, mParticle, Snowplow, Amplitude, Mixpanel, Pendo, etc) or your homemade API.
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:34]:
```suggestion
Codegen works with any analytics tool, whether it's any of the common event logging SDKs (e.g. Google Analytics 4, Segment, RudderStack, mParticle, Snowplow, Amplitude, Mixpanel, PostHog, FreshPaint, Pendo, etc) or an homemade API or SDK.
```

File: pages/implementation/start-using-avo-codegen.mdx [hunk 10]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+
+When designing the tracking plan you configure the list of analytics destinations for each source. In the generated code each destination is represented as its own Destination Interface in the initAvo method. This is how Codegen routes data, locally in your client (without the data ever going through Avo's servers), to the correct destination based on the source/destination connections you've defined in the Avo dashboard.
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:36]: Add a link to destination interface here!

```suggestion
When designing the tracking plan you configure the list of data destinations for each source. In the generated code each destination is represented as its own Destination Interface in the initAvo method. This is how Codegen routes data, locally in your client (without the data ever going through Avo's servers), to the correct destination based on the source/destination connections you've defined in the Avo dashboard.
```

File: pages/implementation/start-using-avo-codegen.mdx [hunk 11]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+
+In the Avo Tracking Plan you can configure which analytics <Link passHref href="/workspace/tracking-plan/events#a-nameactionsa-actions"><a>actions</a></Link> should be included in Codegen (for example you might want to update user properties or log revenue along with logging an event), and Codegen will handle calling all the relevant Destination Interface methods. All you have to do is call the generated function in your code base and pass in type safe key-value pairs, and the Codegen will take care of the rest.
+
+Learn more technical details about Codegen <Link passHref href="/implementation/avo-functions-overview"><a>here</a></Link>.
+
+## Do events flow through Avo servers?
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:42]:
```suggestion
## Does my event data flow through Avo servers?
```

File: pages/implementation/start-using-avo-codegen.mdx [hunk 12]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+
+No. None of your events ever go through Avo servers. They go directly from your devices (whether that's a front-end client or a server) to the analytics destination.
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:44]:
```suggestion
No. None of your events ever go through Avo servers. They go directly from your devices (whether that's a front-end client or a server) to the data destination.
```

File: pages/implementation/start-using-avo-codegen.mdx [hunk 13]
```diff
@@ -5,15 +5,59 @@ title: 'Start using Avo Codegen'
...
+
+> 💡 This sometimes called "device mode". With CDPs, you can choose between <Link passHref href="https://segment.com/docs/guides/intro-impl/#connection-modes"><a>_Cloud-mode_ or _Device-mode_</a></Link>:
+>
+> - In _Cloud-mode_, your events go through the CDP servers. Your devices send the events to the CDP cloud (aka the CDP Servers) and the CDP cloud sends them to the analytics destinations.
+> - In _Device-mode_ the events go directly from your devices (whether it's a front end client or a server) to the analytics destination.
+>   In this regard, Avo works like _Device-mode_.
+
+## How to set up and use Codegen
```
Comment (logason) [inline pages/implementation/start-using-avo-codegen.mdx:52]: I'm wondering if we should split this doc in two somehow.

Feel like everything above is answering the question about what Codegen is, and everything below is about how to get started (quickstart guide)?

**Review by logason — COMMENTED** [review]
Added few comments! I believe we should also gets Rhys and Sang's review here, particularly because they are still getting to know Avo, so it will be a good user test to check if they can follow along 🙌

File: pages/data-design/start-using-inspector.mdx (hunk 2 above)
Comment (hugihlynsson) [reply pages/data-design/start-using-inspector.mdx:81]: Wait, the first sentence says: 

> If you’re not using one of the supported CDPs you can send events data with an SDK.

That reads to me that we are indeed pitching the CDPs as the primary path and the SDK as a fallback. Do you not agree @logason?

File: pages/implementation/devs-101.mdx (hunk 4 above)
Comment (alex-tpom6oh) [reply pages/implementation/devs-101.mdx:1]: Yeah everything from here was moved to the Codegen overview and Inspector overview docs

Comment (alex-tpom6oh) [reply pages/implementation/devs-101.mdx:1]: So maybe we'd need to write a new devs-101 doc

Comment (alex-tpom6oh) [reply pages/implementation/devs-101.mdx]: Brought it back

File: pages/implementation/start-using-avo-codegen.mdx (hunk 13 above)
Comment (alex-tpom6oh) [reply pages/implementation/start-using-avo-codegen.mdx:52]: Yes, I think it makes sense. 👍

**Comment by alex-tpom6oh** [comment]
Applied your suggestions @logason and rearranged the docs a bit. Also made sure that there are documents on the links we used to have before.

**Review by 1212spark — CHANGES_REQUESTED** [review]
Please see comment for row 12 of [explicit-null-in-codegen.mdx](https://github.com/avohq/docs/pull/657/files#diff-04c0121827027f91e0b89e2cef697ef45198fd7c9cfa16fcc76c00e952d3885b). Otherwise reads great!

File: pages/implementation/explicit-null-in-codegen.mdx [hunk 14]
```diff
@@ -9,7 +9,7 @@ With the Avo Codegen our goal is to provide developers with a type safe analytic
 
 One of the ways we do it is with the required values for the `Sometimes Sent` parameters.
 
-The fact that we require to explicitly pass a `null` value for every `Sometimes Sent` property when it's not sent allows the compiler to catch a very common implementation related data bug.
+The fact that we require to explicitly pass a `null` value for every `Sometimes Sent` property when it's not sent allows the compiler to catch a very common implementation related data bugs.
```
Comment (alex-tpom6oh) [inline pages/implementation/explicit-null-in-codegen.mdx:12]:
```suggestion
The fact that we require an explicit null value to be passed for every Sometimes Sent property when it's not sent allows the compiler to catch very common implementation related data bugs.
```

**Review by 1212spark — APPROVED** [review]
thanks @alex-tpom6oh !
